### PR TITLE
Add support for loading initial trajectory pose

### PR DESCRIPTION
**Pull request by damienrg**
#606 add initial trajectory pose to TrajectoryBuilderOptions proto but it was not possible to load it.

### CODE REVIEW

**Comment by damienrg** [comment]
Was Cartographer Open House recorded? If yes is it possible to download it as I think there are valid arguments for going in the command-line flag in cartographer_start_trajectory though I am not sure if it is the best way for now.

**Comment by SirVer** [comment]
No, unfortunately we do not report the open houses. But @cschuet should upload the slides soonish. What other ideas do you have for providing the initial pose?

**Comment by cschuet** [comment]
Slides are up, but there is nothing on there that reflects them.

How about you write a short RFC, Damien, where you explain how think it
should be done and then we can discuss on there?

File: cartographer/mapping_3d/sparse_pose_graph.h [hunk 1]
```diff
@@ -104,9 +104,9 @@ class SparsePoseGraph : public mapping::SparsePoseGraph {
   mapping::MapById<mapping::NodeId, mapping::TrajectoryNode>
   GetTrajectoryNodes() override EXCLUDES(mutex_);
   std::vector<Constraint> constraints() override EXCLUDES(mutex_);
-  void SetInitialTrajectoryPose(int from_trajectory_id, int to_trajectory_id,
-                                const transform::Rigid3d& pose,
-                                const common::Time time) override
+  void SetInitialTrajectoryPose(
```
Comment (SirVer) [inline cartographer/mapping_3d/sparse_pose_graph.h]: I am not sure why this is required. Adding the initial pose through the TrajectoryBuilderOptions should be sufficient, right?

File: configuration_files/trajectory_builder.lua [hunk 2]
```diff
@@ -12,11 +12,13 @@
...
 
+include "initial_trajectory_pose.lua"
 include "trajectory_builder_2d.lua"
 include "trajectory_builder_3d.lua"
 
 TRAJECTORY_BUILDER = {
   trajectory_builder_2d = TRAJECTORY_BUILDER_2D,
   trajectory_builder_3d = TRAJECTORY_BUILDER_3D,
   pure_localization = false,
+  initial_trajectory_pose = INITIAL_TRAJECTORY_POSE,
```
Comment (SirVer) [inline configuration_files/trajectory_builder.lua]: I am not sure this is a good approach. While it is symmetric to how pure_localization is handled, this is quite different in the use case.

Pure localization always goes along with different settings, for example the loop closure will be run more often usually and maybe less aggressive settings for search windows. 

Initial pose however is very dependant on the data set in question: once you have one run finished, you want to start a second and relate the initial pose to this particular state. It sounds more like a flag than a configuration paramater to me. Therefore I feel that a commandline flag would be more appropriate for it.

Wdyt?

File: cartographer/mapping_3d/sparse_pose_graph.h (hunk 1 above)
Comment (damienrg) [reply cartographer/mapping_3d/sparse_pose_graph.h]: TrajectoryBuilderOptions is not available in sparse_pose_graph. Maybe you mean that I should do this https://github.com/googlecartographer/cartographer/pull/606#discussion_r149653974.

File: configuration_files/trajectory_builder.lua (hunk 2 above)
Comment (damienrg) [reply configuration_files/trajectory_builder.lua]: Yes it is cumbersome to do it this way and as you mentioned it was to be consistent with pure localization. Can you elaborate a bit about command line flag? Is it a flag that should be passed to start_trajectory_main in cartographer_ros? If yes, what flag do you have in mind, something like:
```
-initial_pose "{ to_trajectory_id: 1, timestamp: now, translation : { 0, 0, 0}, rotation : { ... }}"
```

**Comment by SirVer** [comment]
> Is it a flag that should be passed to start_trajectory_main in cartographer_ros?

Yes, I think so. That seems to be the correct place to me - after all that is the binary you need to run to start a new trajectory. And the initial pose is a one-time option, so commandline seems right. I just wonder about how it is provided.

> initial_pose "{ to_trajectory_id: 1, timestamp: now, translation : { 0, 0, 0}, rotation : { ... }}"

I think a good way would be to write this as a lua transform, like in [transform.lua](https://github.com/googlecartographer/cartographer_ros/blob/master/cartographer_ros/configuration_files/transform.lua). I.e. in the general case:

    start_trajectory_main --initial_pose '{ translation = { 0., 0., 0. }, rotation = { 0., -math.pi / 2., 0., } }' 

This can be parsed as `LuaParameterDictionary("return " + FLAGS_initial_pose`.

**Comment by cschuet** [comment]
We also need `to_trajectory_id` and `timestamp` so this would look like e.g.

    start_trajectory_main --initial_pose '{to_trajectory_id = 0, timestamp = 1234, relative_pose = { translation = { 0., 0., 0. }, rotation = { 0., -math.pi / 2., 0., } } }'

**Comment by damienrg** [comment]
Closing this and I will open a PR for cartographer_ros.

**Comment by jihoonl** [comment]
@damienrg How's your progress? I would like to take over if you are stuck by something. I got a working version internally.

**Pull request closed without merging**